pkg/driver/node.go

```diff
@@ -110,21 +114,71 @@ func (ns *S3NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePubl
 	}
 
 	klog.V(4).Infof("NodePublishVolume: mounting %s at %s with options %v", bucket, target, mountpointArgs)
+
 	hostPluginDir := os.Getenv(hostPluginDirEnv)
 	if hostPluginDir == "" {
 		// set the default in case the env variable isn't found
 		hostPluginDir = "/var/lib/kubelet/plugins/s3.csi.aws.com/"
 	}
 	hostTokenPath := path.Join(hostPluginDir, "token")
+
+	awsRoleArn := os.Getenv(roleArnEnv)
+
+	usePodIdentity := os.Getenv(podInfoOnMountEnabled)
+	if usePodIdentity == "true" {
+		klog.V(4).Infof("NodePublishVolume: Using pod identity")
+		tokensJson := req.VolumeContext["csi.storage.k8s.io/serviceAccount.tokens"]
+		if tokensJson != "" {
+			var tokens map[string]*Token
+			if err := json.Unmarshal([]byte(tokensJson), &tokens); err != nil {
+				return nil, status.Error(codes.InvalidArgument, "Tokens bad format")
+			}
+			stsToken := tokens["sts.amazonaws.com"]
+			if stsToken != nil {
+				klog.V(4).Infof("NodePublishVolume: stsToken exp: %v", stsToken.ExpirationTimestamp)
+				hostTokenPath = path.Join(hostPluginDir, volumeID+".token")
+				// TODO cleanup these files on unmount and startup
+				os.WriteFile(path.Join("/csi/", volumeID+".token"), []byte(stsToken.Token), 0400)
```

Comment:
Shouldn't we write to `hostTokenPath` here?

Reply:
No - we're writing from inside the CSI Driver container here, which mounts `/csi/` as `{{ trimSuffix "/" .Values.node.kubeletPath }}/plugins/s3.csi.aws.com/` in the underlying host.